src/FlexVotingClient.sol

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity >=0.8.10;
+
+import {SafeCast} from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+import {Checkpoints} from "@openzeppelin/contracts/utils/Checkpoints.sol";
+import {IFractionalGovernor} from "src/interfaces/IFractionalGovernor.sol";
+import {IVotingToken} from "src/interfaces/IVotingToken.sol";
+
+// TODO natspec
+// this is a contract to make it easy to build clients for flex voting governors.
+abstract contract FlexVotingClient {
+  using SafeCast for uint256;
+  using Checkpoints for Checkpoints.History;
+
+  /// @notice The voting options corresponding to those used in the Governor.
+  enum VoteType {
+    Against,
+    For,
+    Abstain
+  }
+
+  /// @notice Data structure to store vote preferences expressed by depositors.
+  struct ProposalVote {
+    uint128 againstVotes;
+    uint128 forVotes;
+    uint128 abstainVotes;
+  }
+
+  /// @notice Map proposalId to an address to whether they have voted on this proposal.
+  mapping(uint256 => mapping(address => bool)) private proposalVotersHasVoted;
+
+  /// @notice Map proposalId to vote totals expressed on this proposal.
+  mapping(uint256 => ProposalVote) public proposalVotes;
+
+  /// @notice The governor contract associated with this governance token. It
+  /// must be one that supports fractional voting, e.g. GovernorCountingFractional.
+  IFractionalGovernor public immutable GOVERNOR;
```

Comment:
Are we sure this can be safely immutable? What if a protocol migrates to a new governor contract and transfers ownership? RR

Reply:
Hm, that's a great point. This probably shouldn't be immutable. The abstract client should also probably require inheriting contracts to define a function to update the governor as well

Since this is a somewhat separate concern, tracking https://github.com/ScopeLift/flexible-voting/issues/48